src/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfo.cs

```diff
@@ -189,6 +189,14 @@ private void DetermineSerializationCapabilities(JsonSerializerOptions options)
 
                 // We deserialize if there is a setter + no [Ignore] attribute. 
                 ShouldDeserialize = HasSetter;
+
+                if (ClassType == ClassType.Dictionary &&
+                    ShouldDeserialize && DefaultImmutableConverter.TypeIsImmutableDictionary(RuntimePropertyType))
+                {
+                    EnumerableConverter = s_jsonImmutableConverter;
+                    ((DefaultImmutableConverter)EnumerableConverter).RegisterImmutableCollectionType(
```

Comment:
You can avoid the cast, and a some confusion if you put L196 and L167  in the opposite order.

When I read this I'm like 🤔 *why is this cast safe?*.